syntax/grammar.mjs

```diff
@@ -0,0 +1,495 @@
+import * as FTL from "./ast.mjs";
+import into from "./abstract.mjs";
+import {
+    after, always, and, between, char, charset, defer, either, eof, maybe, not,
+    regex, repeat, repeat1, sequence, string
+} from "../lib/combinators.mjs";
+import {
+    flatten, mutate, print, to_array, to_object, to_string
+} from "../lib/util.mjs";
+
+/* ------------------------------- */
+/* An FTL file defines a Resource. */
+export
+let Resource = defer(() =>
+    repeat(
+        either(
+            blank_line,
+            Entry.as("entry")))
+    .map(to_array)
+    .map(into(FTL.Resource)));
+
+export
+let Entry = defer(() =>
+    either(
+        Message,
+        Term,
+        either(
+            ResourceComment,
+            GroupComment,
+            Comment),
+        Junk));
+
+let Message = defer(() =>
+    sequence(
+        maybe(Comment).as("comment"),
+        Identifier.as("id"),
+        between(
+            maybe(inline_space),
+            char("=")),
+        either(
+            sequence(
+                Pattern.as("value"),
+                repeat(Attribute).as("attributes")),
+            sequence(
+                always(null).as("value"),
+                repeat1(Attribute).as("attributes"))),
+        line_end)
+    .map(flatten(1))
+    .map(to_object)
+    .map(into(FTL.Message)));
+
+let Term = defer(() =>
+    sequence(
+        maybe(Comment).as("comment"),
+        TermIdentifier.as("id"),
+        between(
+            maybe(inline_space),
+            char("=")),
+        Pattern.as("value"),
+        repeat(Attribute).as("attributes"),
+        line_end)
+    .map(to_object)
+    .map(into(FTL.Term)));
+
+let Attribute = defer(() =>
+    sequence(
+        break_indent,
+        char("."),
+        Identifier.as("id"),
+        between(
+            maybe(inline_space),
+            char("=")),
+        Pattern.as("value"))
+    .map(to_object)
+    .map(into(FTL.Attribute)));
+
+let Comment = defer(() =>
+    repeat1(
+        sequence(
+            char("#"),
+            comment_line.as("line")))
+    .map(flatten(1))
+    .map(to_array)
+    .map(to_string)
+    .map(into(FTL.Comment)));
+
+let GroupComment = defer(() =>
+    repeat1(
+        sequence(
+            string("##"),
+            comment_line.as("line")))
+    .map(flatten(1))
+    .map(to_array)
+    .map(to_string)
+    .map(into(FTL.GroupComment)));
+
+let ResourceComment = defer(() =>
+    repeat1(
+        sequence(
+            string("###"),
+            comment_line.as("line")))
+    .map(flatten(1))
+    .map(to_array)
+    .map(to_string)
+    .map(into(FTL.ResourceComment)));
+
+let Junk = defer(() =>
+    repeat1(
```

Comment:
Looking at how this renders and reads, what do you think about making Junk just single line in the syntax, and glue Junk nodes together in in abstract.mjs?

That would Junk just make "eat the rest of the line", right?

Reply:
Interesting idea, that would work, yes. It would only avoid the `repeat1` though, right? I think I'd prefer to keep it as it is right now. In fact, I already have a few ideas for changes that will reduce the amount of post-processing in `abstract.mjs`. If something can be expressed in the grammar, my current thinking is to do it there.

Reply:
I expect it to actually remove all the needs for `not(Entry)`

Reply:
That would be a win. I'll try it, thanks!

Reply:
I tried it out and it works well except that `Junk` doesn't now include trailing blank lines (if any). Which might actually make it a bit more like other `Entries`.

Reply:
so if we'd join them afterwards, we'd have to pad our own newline? Or add top-level whitespace to the `Resource` production?

Reply:
Post-processing Junk changes how the following is parsed:

```
junk_line1\n
\n
junk_line2\n
\n
```

Before it was a single `Junk` entry which included all 4 `\n`. When we join junk lines in post-processing, those blank_lines don't make it there because they are dropped from the CST before `map(into(FTL.Resource))` is called. In a naive joining approach, the new single `Junk` would now only contain `junk_line1\njunk_line2`. Note the missing middle and trailing `\n`

One option is to let blank lines make it to the `FTL.Resource` construction and strip them there as the last post-processing step. This prevents the two junk lines above from being joined because there aren't adjacent now. The result is that we end up with _two_ `Junk` entries now, each for a contiguous block of junk lines.

I implemented this behavior in a commit which I'll push in a few.